src/miral/basic_window_manager.cpp

```diff
@@ -564,6 +564,56 @@ void miral::BasicWindowManager::focus_next_application()
     select_active_window(focussed_surface ? info_for(focussed_surface).window() : Window{});
 }
 
+void miral::BasicWindowManager::focus_prev_application()
+{
+    if (auto const prev = active_window())
+    {
+        auto const workspaces_containing_window = workspaces_containing(prev);
+
+        if (!workspaces_containing_window.empty())
+        {
+            do
+            {
+                focus_controller->focus_prev_session();
+
+                if (can_activate_window_for_session_in_workspace(
+                    focus_controller->focused_session(),
+                    workspaces_containing_window))
+                {
+                    return;
+                }
+            }
+            while (focus_controller->focused_session() != prev.application());
+        }
+        else
+        {
+            do
+            {
+                focus_controller->focus_prev_session();
+
+                if (can_activate_window_for_session(focus_controller->focused_session()))
+                {
+                    return;
+                }
+            }
+            while (focus_controller->focused_session() != prev.application());
+        }
+    }
+    else
+    {
+        focus_controller->focus_prev_session();
+
+        if (can_activate_window_for_session(focus_controller->focused_session()))
+        {
+            return;
+        }
+    }
+
+    // Last resort: accept wherever focus_controller places focus
+    auto const focussed_surface = focus_controller->focused_surface();
+    select_active_window(focussed_surface ? info_for(focussed_surface).window() : Window{});
+}
```

Comment:
Looks like there's some code duplication in this function with `focus_next_application()`. Would it make sense to merge the two? (perhaps keep the names for ABI compat but implement them each as just a call to a new function: `switch_focus_application(bool)`)

Reply:
Well, we _could_ pass a `void FocusController::(*focus_some_session)()` pointer into a shared implementation function, but I don't think we'd gain much clarity. (Or use a template I suppose.)

Reply:
There seems to be a *reasonable* amount of logic in these, and it's trivially duplicated; it would be *nice* to consolidate that.

I won't block on that, but maybe it's something to explore?